Retry to synchronize block at height when error occurs

Simple error during synchronization process can now cause inconsistency in explorer db:

```
15:34:58.235 ERROR grabber-service - Error while requesting block with id dbe37a2d28ccd6f44229faf50225e013b85b7f093635eddc9486acbc327ab9f2 
java.net.SocketTimeoutException: Read timed out
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1944)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1939)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.net.www.protocol.http.HttpURLConnection.getChainedException(HttpURLConnection.java:1938)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1508)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
	at scalaj.http.HttpRequest.doConnection(Http.scala:367)
	at scalaj.http.HttpRequest.exec(Http.scala:343)
	at org.ergoplatform.explorer.grabber.http.RequestServiceImpl.executeRequest(RequestService.scala:64)
	at org.ergoplatform.explorer.grabber.http.RequestServiceImpl.$anonfun$getIO$1(RequestService.scala:27)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:128)
	at cats.effect.internals.IORunLoop$.startCancelable(IORunLoop.scala:42)
	at cats.effect.internals.IOParMap$$anon$1.run(IOParMap.scala:57)
	at cats.effect.internals.TrampolineEC.cats$effect$internals$TrampolineEC$$localRunLoop(TrampolineEC.scala:62)
	at cats.effect.internals.TrampolineEC.$anonfun$execute$1(TrampolineEC.scala:54)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
	at cats.effect.internals.TrampolineEC.execute(TrampolineEC.scala:54)
	at cats.effect.internals.Callback$AsyncIdempotentCallback.apply(Callback.scala:139)
	at cats.effect.internals.Callback$AsyncIdempotentCallback.apply(Callback.scala:130)
	at cats.effect.internals.IOTimer$Tick.run(IOTimer.scala:88)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
	at sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:735)
	at sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:678)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1587)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
	at scalaj.http.HttpRequest.doConnection(Http.scala:365)
	... 17 common frames omitted
15:34:58.240 INFO  grabber-service - 0 block(s) from height 415 has been written to db 
15:34:58.922 ERROR grabber-service - An error has occurred:  
doobie.util.invariant$UnexpectedEnd$: ResultSet exhausted; more rows expected.
	at doobie.util.invariant$UnexpectedEnd$.<clinit>(invariant.scala)
	at doobie.hi.resultset$.$anonfun$getUnique$1(resultset.scala:195)
	at doobie.hi.resultset$.$anonfun$getUnique$1$adapted(resultset.scala:192)
	at cats.SemigroupalArityFunctions.$anonfun$map2$1(SemigroupalArityFunctions.scala:30)
	at cats.Monad.$anonfun$map$1(Monad.scala:16)
	at cats.free.Free.$anonfun$step$1(Free.scala:52)
	at cats.free.Free.step(Free.scala:53)
	at cats.free.Free.$anonfun$foldMap$1(Free.scala:153)
	at cats.data.KleisliFlatMap.$anonfun$tailRecM$2(Kleisli.scala:405)
	at cats.effect.IOInstances$$anon$1.tailRecM(IO.scala:488)
	at cats.effect.IOInstances$$anon$1.$anonfun$tailRecM$1(IO.scala:489)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:128)
	at cats.effect.internals.IORunLoop$.startCancelable(IORunLoop.scala:42)
	at cats.effect.internals.IOParMap$$anon$1.run(IOParMap.scala:57)
	at cats.effect.internals.TrampolineEC.cats$effect$internals$TrampolineEC$$localRunLoop(TrampolineEC.scala:62)
	at cats.effect.internals.TrampolineEC.$anonfun$execute$1(TrampolineEC.scala:54)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
	at cats.effect.internals.TrampolineEC.execute(TrampolineEC.scala:54)
	at cats.effect.internals.Callback$AsyncIdempotentCallback.apply(Callback.scala:139)
	at cats.effect.internals.Callback$AsyncIdempotentCallback.apply(Callback.scala:130)
	at cats.effect.internals.IOTimer$Tick.run(IOTimer.scala:88)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/src/main/scala/org/ergoplatform/explorer/grabber/GrabberService.scala
+++ b/src/main/scala/org/ergoplatform/explorer/grabber/GrabberService.scala
@@ -30,7 +30,7 @@ class GrabberService(xa: Transactor[IO], executionContext: ExecutionContext, con
 
   private val active = new AtomicBoolean(false)
   private val pause = config.grabber.pollDelay
-
+  private val MaxRetriesNumber = 4
 
   private def idsAtHeight(height: Long): IO[List[String]] =
     requestService.get[List[String]](addressService.idsAtHeightUri(height))
@@ -51,7 +51,7 @@ class GrabberService(xa: Transactor[IO], executionContext: ExecutionContext, con
     .map(_.flatten.toList)
     .map(_.zipWithIndex.map { case (bs, i) => bs -> (i == 0)})
 
-  def writeBlocksFromHeight(h: Long): IO[Unit] = {
+  def writeBlocksFromHeight(h: Long, retries: Int = 0): IO[Unit] = {
     def updatedBlock(b: ApiFullBlock, isMain: Boolean) = b.copy(header = b.header.copy(mainChain = isMain))
     for {
       ids <- idsAtHeight(h)
@@ -62,9 +62,14 @@ class GrabberService(xa: Transactor[IO], executionContext: ExecutionContext, con
       _ <- blocks.map { case (block, isMain) =>
         writeBlockInfo(updatedBlock(block, isMain))
       }.parSequence
-      _ <- IO {
-        logger.info(s"${blocks.length} block(s) from height $h has been written to db")
+      retryNeeded <- IO {
+        blocks.isEmpty && retries < MaxRetriesNumber
       }
+      _ <- IO {
+        val retryInfo = if (retryNeeded) "Retrying.." else ""
+        logger.info(s"${blocks.length} block(s) from height $h has been written. $retryInfo")
+      }
+      _ <- if (retryNeeded) writeBlocksFromHeight(h, retries + 1) else IO.unit
     } yield ()
   }
 
@@ -106,7 +111,7 @@ class GrabberService(xa: Transactor[IO], executionContext: ExecutionContext, con
         Range.Long.inclusive(currentHeight + 1, info.fullHeight, 1L).toList
       }
     }
-    _ <- heightsRange.map { writeBlocksFromHeight }.sequence[IO, Unit]
+    _ <- heightsRange.map(writeBlocksFromHeight(_)).sequence[IO, Unit]
     _ <- IO {
       logger.info(s"Sync task has been finished. Current height now is ${info.fullHeight}.")
     }

--- a/src/main/scala/org/ergoplatform/explorer/grabber/db/DBHelper.scala
+++ b/src/main/scala/org/ergoplatform/explorer/grabber/db/DBHelper.scala
@@ -80,6 +80,6 @@ class DBHelper(networkConfig: NetworkConfig) {
   }
 
   def readCurrentHeight: ConnectionIO[Long] =
-    fr"SELECT COALESCE(height, CAST(0 as BIGINT)) FROM node_headers ORDER BY height DESC LIMIT 1"
+    fr"SELECT COALESCE(height, CAST(0 as BIGINT)) FROM blocks_info ORDER BY height DESC LIMIT 1"
       .query[Long].option.map { _.getOrElse(Constants.PreGenesisHeight) }
 }

